### PR TITLE
Forbid access to deleted assessments and questions

### LINKED ISSUE #7505: Forbid access to deleted assessments
If someone navigates to any of the instructor assessment pages by ID, we don't check for `deleted_at IS NULL` before loading the page. This can lead to other errors as various files or data that we expect to be present can't be found. We should 404 for deleted assessments.

### PR DESCRIPTION
**Pull request by nwalters512**
Closes #7505.

It's possible that more of our `selectAndAuthz*` middlewares need something similar? I spot-checked and it seems like they're mostly doing the right thing. A few I'm not sure about:

- `selectAndAuthzInstanceQuestion`: should this require `q.deleted_at IS NULL`?
- `selectAndAuthzAssessmentQuestion` - should this require `q.deleted_at IS NULL AND aq.deleted_at IS NULL`?

### CODE REVIEW

**Comment by mwest1066** [comment]
> * `selectAndAuthzInstanceQuestion`: should this require `q.deleted_at IS NULL`?
> * `selectAndAuthzAssessmentQuestion` - should this require `q.deleted_at IS NULL AND aq.deleted_at IS NULL`?

Yes to all of these, I think.

File: middlewares/selectAndAuthzInstanceQuestion.sql [hunk 1]
```diff
@@ -155,5 +155,6 @@ WHERE
     $assessment_id::BIGINT IS NULL
     OR a.id = $assessment_id
   )
+  AND q.deleted_at IS NULL
```
Comment (mwest1066) [inline middlewares/selectAndAuthzInstanceQuestion.sql:158]: Do we also want checks on `aq.deleted_at` and `a.deleted_at` here? Conceivably also `c.deleted_at`? I'm not sure how high up the chain it's worth enforcing?

Comment (nwalters512) [reply middlewares/selectAndAuthzInstanceQuestion.sql:158]: The first two make sense! I think `c.deleted_at` will (should) always be caught by `middlewares/authzCourseOrInstance`?

Comment (nwalters512) [reply middlewares/selectAndAuthzInstanceQuestion.sql:158]: I added checks for `aq.deleted_at` and `a.deleted_at` in b0c6979.

Comment (mwest1066) [reply middlewares/selectAndAuthzInstanceQuestion.sql:158]: Good point on course access being checked already.

**Review by mwest1066 — APPROVED** [review]

File: middlewares/selectAndAuthzInstanceQuestion.sql (hunk 1 above)
Comment (jonatanschroeder) [reply middlewares/selectAndAuthzInstanceQuestion.sql:158]: Is there a problem here with using `aq.deleted_at`? What happens if a student opens a question that was active when they took the assessment, but deleted afterwards?

Comment (nwalters512) [reply middlewares/selectAndAuthzInstanceQuestion.sql:158]: We check for `aq.deleted_at` in many places already (grep for `aq.deleted_at IS NULL`), so this is just making it consistent with everywhere else. If an assessment question is deleted, I think the correct thing to do is to not show it to students and forbid access to it even if they manage to guess the URL. Does that sound right to you?

Comment (jonatanschroeder) [reply middlewares/selectAndAuthzInstanceQuestion.sql:158]: That makes sense. I didn't check the list, I was just confirming that this is the intended behaviour, particularly for exams, since they don't regenerate instance questions when the assessment changes.

Comment (mwest1066) [reply middlewares/selectAndAuthzInstanceQuestion.sql:158]: This has been a sore point for a long time, actually. What happens is that we are running an exam in the CBTF, the instructor realizes on day 2 that one question in an `alternative_group` is partially broken, and they want to remove this question from the assessment so that future students won't be assigned it. However, they want existing exams taken on day 1 to still show the question that the student did, especially because the question might have worked fine for these students (and we want to show how their score was computed).

I found the old issue about this: https://github.com/PrairieLearn/PrairieLearn/issues/1356

In summary, I think we want to allow access to instance questions even though the assessment question has been deleted. However, handling this cleanly will probably require some more code changes beyond just changing `deleted_at` filters, because we'll need to track the `alternative_group` and `zone` information that might have been deleted.

Comment (mwest1066) [reply middlewares/selectAndAuthzInstanceQuestion.sql:158]: Yes, let's _not_ add `aq.deleted_at IS NULL` here.

Comment (nwalters512) [reply middlewares/selectAndAuthzInstanceQuestion.sql]: Removed that check in fd5f67e.

Comment (nwalters512) [reply middlewares/selectAndAuthzInstanceQuestion.sql]: Whoops, there was one more that I removed in 2815b39.

**Review by mwest1066 — APPROVED** [review]
